src/modules/pipeline.py

```diff
@@ -1,10 +1,183 @@
 
-class Pipeline():
+        self.test_loader = DataLoader(dataset.test_set, batch_size=128)
+        self.train_loader = DataLoader(dataset.labeled_set, batch_size=32, shuffle=True)
 
-    # TODO What parameters
+        if type(settings) == dict:
+            settings = PipelineSettings.from_dict(settings)
+        self.settings: PipelineSettings = settings
 
-    pass
\ No newline at end of file
+    def run(self) -> tuple[em.Session, dict[str, list[float]]]:
+        """
+        Runs the pipeline. Return a Session object as well as a dict of stats, which contains
+        loss history on train and test and accuracy history
+
+        Pseudocode:
+
+        perform initial learner fit
+        loop for n_queries
+              get_queries
+              label_queries
+              fit learner on updated train set
+              update session
+        return session
+        """
+        # initialise loss/accuracy history arrays
+        # initial learner fit
+        t_loss_history, v_loss_history = self.learner.fit(
+            self.train_loader, self.test_loader,
+            self.optimiser, self.loss_function,
+            epochs=self.settings.init_epochs,
+            early_stopping=True,
+        )
+        accuracy_history = [self._calculate_accuracy(self.test_loader).item()]
+
+        # initialise the session
+        session = em.Session(dataset=self.dataset, learner=self.learner)
+
+        for i in range(self.settings.n_queries):
+            print(f"Iteration: {i+1}")
+
+            # generate queries
+            unlabeled_loader = DataLoader(self.dataset.unlabeled_set, batch_size=128, shuffle=False)
+            queries, uncertainty = self.learner.generate_query(
+                unlabeled_loader,
+                criterion='entropy',
+                batch_size=self.settings.query_batch_size,
+            )
+            # label queries
+            self.dataset.get_label_by_idx(queries, move_sample=True)
+            print(f"Max uncertainty: {uncertainty}")
+            self.train_loader = DataLoader(self.dataset.labeled_set, batch_size=32, shuffle=True)
+
+            # fit learner on updated train set
+            t_hist, v_hist = self.learner.fit(
+                self.train_loader, self.test_loader,
+                self.optimiser, self.loss_function,
+                epochs=self.settings.epochs_per_query,
+                early_stopping=True,
+                sample_weights=None,
+            )
+            # update session
+            session.update()
+
+            # update loss/accuracy history
+            t_loss_history.extend(t_hist)
+            v_loss_history.extend(v_hist)
+            accuracy_history.append(self._calculate_accuracy(self.test_loader).item())
+
+        # pack loss/accuracy history arrays into one dict
+        stats_to_return = {
+            't_loss_history': t_loss_history,
+            'v_loss_history': v_loss_history,
+            'accuracy_history': accuracy_history,
+        }
```

Comment:
Add `f1_score_history` as one of the returned stats. Add parameters to `run` function:
- `calculate_accuracy=False`
    - only calculate accuracy if True (to speed up computation)
- `calculate_f1_score=False`
    - same as above